### PR TITLE
FiberCacheManager toDebugString and CacheStats class

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

A method to return description of this FiberCacheManager for debugging.
New class called CacheStats to avoid depending on Guava

## How was this patch tested?

Unit test pass

### CODE REVIEW

**Comment by linhong-intel** [comment]
Output looks like:
FiberCacheManager Statistics: { cacheCount=63, usedMemory=63.0 KB, hitCount=126, missCount=126, loadCount=126, totalLoadTime=7037718ns, evictionCount=63 }

**Comment by linhong-intel** [comment]
Fixes #436

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * Note: But for now, we don't need to record the stats since Guava has done it for us.
+ *
+ * @param hitCount incremented when a cache lookup encounters an existing cache entry
+ * @param missCount incremented when a cache lookup first encounters a missing cache entry
+ * @param loadCount incremented when a new entry is loaded
+ * @param totalLoadTime incremented when a new entry is loaded
+ * @param evictionCount incremented when an entry is evicted from the cache
+ */
+case class CacheStats(
+    hitCount: Long, missCount: Long, loadCount: Long, totalLoadTime: Long, evictionCount: Long) {
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala]: nit: can we use the below style to be consistent with most Spark code?
```
case class CacheStats(
    hitCount: Long,
    missCount: Long,
    loadCount: Long,
    totalLoadTime: Long,
    evictionCount: Long) {
```

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    val rc = requestCount
+    if (rc == 0) 1.0 else hitCount.toDouble / rc
+  }
+
+  def missRate: Double = {
+    val rc = requestCount
+    if (rc == 0) 0.0 else missCount.toDouble / rc
+  }
+
+  def averageLoadPenalty: Double = if (loadCount == 0) 0.0 else totalLoadTime / loadCount
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala]: `totalLoadTime.toDouble / loadCount`?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    CacheStats(
+      math.max(0, hitCount - other.hitCount),
+      math.max(0, missCount - other.missCount),
+      math.max(0, loadCount - other.loadCount),
+      math.max(0, totalLoadTime - other.totalLoadTime),
+      math.max(0, evictionCount - other.evictionCount))
+}
+
+object CacheStats {
+  def apply(guavaCacheStats: com.google.common.cache.CacheStats): CacheStats =
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala]: Can we remove this? It seems not necessary to involve guava in this class. We can just use CacheStats(guavaCacheStats.hitCount(), guavaCacheStats.missCount(), ...) to create a CacheStats.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 4]
```diff
@@ -99,31 +101,33 @@ object FiberCacheManager extends Logging {
     }
   }
 
+  // NOTE: all members' init should be placed before this line.
+  logDebug(s"Initialized FiberCacheManager")
+
   def get(fiber: Fiber, conf: Configuration): FiberCache = synchronized {
+    logDebug(s"Getting Fiber: $fiber")
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Will this be too noisy?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 5]
```diff
@@ -146,6 +150,19 @@ object FiberCacheManager extends Logging {
 
   // Used by test suite
   private[filecache] def pendingSize: Int = cacheBackend.pendingSize
+
+  // A description of this FiberCacheManager for debugging.
+  def toDebugString: String = {
+    "FiberCacheManager Statistics: { " +
+        s"cacheCount=${cacheBackend.cacheCount}, " +
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: nit: Indent

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala [hunk 6]
```diff
@@ -107,8 +107,10 @@ class GuavaOapCache(cacheMemory: Long, cacheGuardianMemory: Long) extends OapCac
   private def cacheLoader(fiber: Fiber, configuration: Configuration) =
     new Callable[FiberCache] {
       override def call(): FiberCache = {
-        logDebug(s"Loading Cache: $fiber")
+        val startLoadingTime = System.currentTimeMillis()
         val fiberCache = fiber.fiber2Data(configuration)
+        logDebug("Load missed fiber took %s. Fiber: %s"
+            .format(Utils.getUsedTimeMs(startLoadingTime), fiber))
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala]: Nit: Indent. 
Will this also be too noisy in debug mode? Or we can measure it in the perf/benchmark module by calculating the total time of loading a few fibers?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+package org.apache.spark.sql.execution.datasources.oap.filecache
+
+import com.google.common.cache.AbstractCache.StatsCounter
+
+/**
+ * Statistics about the performance of [[OapCache]]. Instances of this class are immutable.
+ * Need another class like [[StatsCounter]] to record stats and return current [[CacheStats]].
+ * Note: But for now, we don't need to record the stats since Guava has done it for us.
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/CacheStats.scala]: Nit: Can we simplify this and not mention anything about guava and its class?

**Comment by carsonwang** [comment]
Good idea to have our own `CacheStats`. Thanks @linhong-intel and I just left some minor comments.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 4 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: How about change to logTrace().
It's good to have some logs to debug customer issues when we can't reproduce it locally.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 8]
```diff
@@ -146,6 +150,19 @@ object FiberCacheManager extends Logging {
 
   // Used by test suite
   private[filecache] def pendingSize: Int = cacheBackend.pendingSize
+
+  // A description of this FiberCacheManager for debugging.
+  def toDebugString: String = {
+    "FiberCacheManager Statistics: { " +
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Nit: Maybe you can use
```scala
s"""
 FiberCacheManager Statistics: { 
 cacheCount=${cacheBackend.cacheCount},
 ...
"""
```
For formatted output string.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 9]
```diff
@@ -42,13 +43,13 @@ private[filecache] class CacheGuardian(maxMemory: Long) extends Thread with Logg
 
   private val _pendingFiberSize: AtomicLong = new AtomicLong(0)
 
-  private val removalPendingQueue = new LinkedBlockingQueue[FiberCache]()
+  private val removalPendingQueue = new LinkedBlockingQueue[(Fiber, FiberCache)]()
 
   def pendingSize: Int = removalPendingQueue.size()
 
-  def addRemovalFiber(fiberCache: FiberCache): Unit = {
+  def addRemovalFiber(fiber: Fiber, fiberCache: FiberCache): Unit = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: From the interface change here, I am a little bit worry as the `fiber` and `fiberCache` is de-coupled, but ideally, they should be coupled. Let's see what's the most elegant way of doing this.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 10]
```diff
@@ -146,6 +150,18 @@ object FiberCacheManager extends Logging {
 
   // Used by test suite
   private[filecache] def pendingSize: Int = cacheBackend.pendingSize
+
+  // A description of this FiberCacheManager for debugging.
+  def toDebugString: String = {
+    s"""FiberCacheManager Statistics: {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: I am wondering if we can move this code into `CacheStats`, as the end user may also want to get the detailed debugging string.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 9 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: sure. I'll think about this.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 10 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Then the output will be something like:
```
FiberCacheManager.toDebugString:
  "FiberCacheManager Statistics { cacheCount: 1, memoryUsed: 2mb, CacheStats: {hit: 1, miss:2, ....} }"
cacheStats.toDebugString:
  "CacheStats: { hit: 1, miss: 2 }"
```

**Comment by linhong-intel** [comment]
@carsonwang @chenghao-intel 
Let's merge this PR?

**Comment by chenghao-intel** [comment]
LGTM, I've merged this @linhong-intel.

**Pull request merged**